gpflow/kernels/categorical.py

```diff
@@ -0,0 +1,97 @@
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from .. import default_int, set_trainable
+from ..base import Parameter, TensorType
+from . import Kernel
+
+tfd = tfp.distributions
+st = tfp.bijectors.Softplus().inverse
+
+
+def latent_from_labels(Z: tf.Tensor, labels: TensorType) -> TensorType:
+    """
+    Converts a tensor of labels into a tensor of the latent space values of those labels.
+    :param Z: Z is a tensor whose entries are the latent space values of the relevant label
+        i.e. Z[3] is the value of label 3.
+        [num_labels]
+    :param X: X is the input data.  The last column should contain the labels as integers.
+        [..., N, D+1]
+    :return: [..., N, D + label_dim]
+    """
+    indices = tf.cast(labels, default_int())  # [..., N]
+    return tf.gather(Z, indices)  # [..., N, label_dim]
+
+
+def _concat_inputs_with_latents(Z: tf.Tensor, X: TensorType) -> TensorType:
+    """
+    Transforms a dataset X from category space into the latent space.
+    :param Z: Z is a tensor whose entries are the latent space values of the relevant label
+        i.e. Z[3] is the value of label 3.
+        [num_labels]
+    :param X: X is the input data.  The last column should contain the labels as integers.
+        [..., N, D+1]
+    :return: [..., N, D + label_dim]
+    """
+    labels = X[..., -1]
+    latent_values = latent_from_labels(Z, labels)
+
+    return tf.concat([X[..., :-1], latent_values], axis=-1)  # [..., N, x_dim + label_dim]
+
+
+class Categorical(Kernel):
+    """
+    A class that implements a categorical latent space wrapper for other Kernels.  It works
+     by dynamically replacing integer labels in the input data with values in a latent space (Z),
+     wrapping two kernels which are combined multiplicatively and otherwise work as normal.
+    Kernel 1 is any kernel of your choice which will deal with the non-categorical dimensions.
+    Kernel 2 deals with the categorical dimension.  It is fixed in training, in order to reduce the number
+     of degrees of freedom in optimisation.  You may also find fixing the lengthscale to be useful here.
+     Note that Z is parameterised by the differences of latent space values for each category for the same reason.
+    Multiple categories can be included by wrapping multiple layers of CategoricalKernel.
+    :param wrapped_kernel_1: The non-categorical kernel.
+    :param wrapped_kernel_2: The categorical kernel.  Set its active_dims to the label dimension.
+    :param num_labels: The number of labels
+    """
+
+    def __init__(
+        self,
+        wrapped_kernel_1: Kernel,
+        wrapped_kernel_2: Kernel,
+        num_labels: int,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        # wrapped_kernel_2 is the one that trains the categorical variables
+        set_trainable(wrapped_kernel_2, False)
+        self.wrapped_kernel = wrapped_kernel_1 * wrapped_kernel_2
+        label_dim = 1
```

Comment:
I thought the categorical_kernel was supposed to specify the label dimension in its active_dims?

Reply:
We only support one label dimension for now, this could be extended at a later date.